### PR TITLE
Use new `dealii::MatrixFreeTools` functions for diagonal computation and matrix assembly

### PR DESCRIPTION
**Pull request by peterrum**
uses https://github.com/dealii/dealii/pull/17029

### CODE REVIEW

File: include/exadg/operators/operator_base.cpp [hunk 1]
```diff
@@ -1302,155 +1295,58 @@ OperatorBase<dim, Number, n_components>::face_loop_empty(
...
-  Range const &                           range) const
+OperatorBase<dim, Number, n_components>::cell_compute_diagonal_matrix(
+  IntegratorCell & integrator) const
 {
-  (void)src;
-
-  IntegratorCell integrator =
-    IntegratorCell(matrix_free, this->data.dof_index, this->data.quad_index);
+  // TODO: this is currently done for every column, but would only be necessary
+  // once per cell
```
Comment (nfehn) [inline include/exadg/operators/operator_base.cpp:1302]: Just to be sure: Is this comment still correct?

**Comment by nfehn** [comment]
Looks good as far as I can judge this, but I am certainly not an expert here. @kronbichler could you have a look at this PR as well?

**Comment by nfehn** [comment]
@peterrum Is this ready to be merged from your side?

**Comment by peterrum** [comment]
> @peterrum Is this ready to be merged from your side?

Yes

File: include/exadg/operators/operator_base.cpp [hunk 2]
```diff
@@ -822,26 +813,28 @@ void
...
-                      &This::boundary_face_loop_calculate_system_matrix,
-                      this,
-                      system_matrix,
-                      system_matrix);
+    dealii::MatrixFreeTools::compute_matrix(*matrix_free,
+                                            matrix_free->get_affine_constraints(),
+                                            system_matrix,
+                                            &This::cell_compute_diagonal_matrix,
+                                            &This::face_compute_diagonal_matrix,
+                                            &This::boundary_face_compute_diagonal_matrix,
```
Comment (nfehn) [inline include/exadg/operators/operator_base.cpp:823]: Here we want to compute the full matrix, not just the diagonal. So I assume this is a bug here?

Comment (nfehn) [reply include/exadg/operators/operator_base.cpp:823]: Ah, I think I am just irritated by the name. Nevertheless, it seems as if the code does not work correctly and I get a matrix with `n_nonzero_elements() = 0`.

**Comment by nfehn** [comment]
Should we revert this PR due to #702?

**Pull request merged**